ihmc-communication/src/main/java/us/ihmc/communication/ros2/sync/ROS2PeerClockOffsetEstimator.java

```diff
@@ -0,0 +1,143 @@
+package us.ihmc.communication.ros2.sync;
+
+import ihmc_common_msgs.msg.dds.PeerClockOffsetEstimatorPingMessage;
+import us.ihmc.commons.exception.DefaultExceptionHandler;
+import us.ihmc.commons.thread.RepeatingTaskThread;
+import us.ihmc.commons.thread.ThreadTools;
+import us.ihmc.communication.ROS2Tools;
+import us.ihmc.communication.packets.MessageTools;
+import us.ihmc.pubsub.common.Guid;
+import us.ihmc.pubsub.common.SampleInfo;
+import us.ihmc.ros2.ROS2Node;
+import us.ihmc.ros2.ROS2Publisher;
+import us.ihmc.ros2.ROS2Topic;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Sends out pings to other instances of this class on the network to estimate
+ * the clock offsets on those computers. This can be used to synchronize data
+ * modifications. We account for and assume a symmetric network delay.
+ */
+public class ROS2PeerClockOffsetEstimator
+{
+   private static final ROS2Topic<PeerClockOffsetEstimatorPingMessage> TOPIC = ROS2Tools.IHMC_ROOT.withModule("peer_clock_offset_estimator")
+                                                                                                  .withType(PeerClockOffsetEstimatorPingMessage.class);
+
+   private final HashMap<Guid, ROS2PeerClockOffsetEstimatorPeer> peerMap = new HashMap<>();
+   private final List<ROS2PeerClockOffsetEstimatorPeer> peerList = new ArrayList<>();
+   private int nextPeerToPing = 0;
+   private final ROS2Publisher<PeerClockOffsetEstimatorPingMessage> publisher;
+   private final Guid ourGuid;
+   private final RepeatingTaskThread requestThread = new RepeatingTaskThread(getClass().getSimpleName(),
+                                                                             this::requestThread,
+                                                                             DefaultExceptionHandler.MESSAGE_AND_STACKTRACE);
+   private final PeerClockOffsetEstimatorPingMessage requestMessage = new PeerClockOffsetEstimatorPingMessage();
+   private final PeerClockOffsetEstimatorPingMessage receivedMessage = new PeerClockOffsetEstimatorPingMessage();
+   private final SampleInfo sampleInfo = new SampleInfo();
+   private final Guid receivedRequestTarget = new Guid();
+   private final Guid receivedReplyTarget = new Guid();
+
+   public ROS2PeerClockOffsetEstimator(ROS2Node ros2Node)
+   {
+      publisher = ros2Node.createPublisher(TOPIC);
+      ourGuid = publisher.getPublisher().getGuid();
```

Comment:
There's a build issue. The `getPublisher()` method of `ROS2Publisher` is package private, so it can't be called here :(

Reply:
Yes it depends on a new release of ros2-library

Reply:
We shouldn't merge until that release